authenticatorbridge/src/main/java/com/bitwarden/authenticatorbridge/manager/AuthenticatorBridgeManagerImpl.kt

```diff
@@ -0,0 +1,189 @@
+package com.bitwarden.authenticatorbridge.manager
+
+import android.content.ComponentName
+import android.content.Context
+import android.content.Intent
+import android.content.ServiceConnection
+import android.os.IBinder
+import androidx.lifecycle.DefaultLifecycleObserver
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.ProcessLifecycleOwner
+import com.bitwarden.authenticatorbridge.IAuthenticatorBridgeService
+import com.bitwarden.authenticatorbridge.manager.model.AccountSyncState
+import com.bitwarden.authenticatorbridge.manager.model.AuthenticatorBridgeConnectionType
+import com.bitwarden.authenticatorbridge.manager.model.toPackageName
+import com.bitwarden.authenticatorbridge.model.EncryptedSharedAccountData
+import com.bitwarden.authenticatorbridge.provider.AuthenticatorBridgeCallbackProvider
+import com.bitwarden.authenticatorbridge.provider.StubAuthenticatorBridgeCallbackProvider
+import com.bitwarden.authenticatorbridge.provider.SymmetricKeyStorageProvider
+import com.bitwarden.authenticatorbridge.util.decrypt
+import com.bitwarden.authenticatorbridge.util.toFingerprint
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+
+private const val AUTHENTICATOR_BRIDGE_SERVICE_CLASS =
+    "com.x8bit.bitwarden.data.platform.service.AuthenticatorBridgeService"
+
+/**
+ * Default implementation of [AuthenticatorBridgeManager].
+ *
+ * @param applicationContext The Context that will be used to bind to AuthenticatorBridgeService. Should be an
+ * application context.
+ * @param connectionType Specifies which build variant to connect to.
+ * @param symmetricKeyStorageProvider Provides access to local storage of the symmetric encryption
+ * key.
+ * @param callbackProvider Provides a way to construct a service callback that can be mocked in
+ * tests.
+ * @param processLifecycleOwner Lifecycle owner that is used to listen for start/stop
+ * lifecycle events.
+ */
+class AuthenticatorBridgeManagerImpl(
+    private val applicationContext: Context,
+    private val connectionType: AuthenticatorBridgeConnectionType,
+    private val symmetricKeyStorageProvider: SymmetricKeyStorageProvider,
+    callbackProvider: AuthenticatorBridgeCallbackProvider = StubAuthenticatorBridgeCallbackProvider(),
+    processLifecycleOwner: LifecycleOwner = ProcessLifecycleOwner.get(),
+) : AuthenticatorBridgeManager, DefaultLifecycleObserver {
+
+    /**
+     * Keep track of whether the service is bound. Useful for making sure we don't call
+     * [Context.unbindService] on a service that isn't bound.
```

Comment:
Would that be a problem?

Reply:
It would throw an exception.

Reply:
What exception?

Reply:
To give more context, I'm not sure how much I trust that boolean; it makes me worry about threading issues and it getting out of sync somehow and that if we need to protect ourselves then we just protect ourselves more directly.

Reply:
Here's the exception that is thrown if you try to unbind a service that isn't bound: `java.lang.IllegalArgumentException: Service not registered`.

The idea for keeping track of `isBound` boolean came from the Android service docs. 

I see your point, though. There is some way this could get out of sync that I don't understand or am not thinking of. We could wrap our `Context.unbindService` call in `try`/`catch`, as a different way of protecting against this case.  

Do you have a preference either way?

Reply:
Well I suppose the one thing to check is if `isBound` value is always getting set on the same thread (likely the main thread). If it is, then when you have is probably fine. Otherwise I think we may need to ditch it and do the try/catch. Again, for this particular feature we have to be super careful about anything that could throw an exception.